synthesizer/benches/add_next_block.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// The snarkVM library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkVM library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg(feature = "testing")]
+
+#[macro_use]
+extern crate criterion;
+
+mod benchmarks;
+use benchmarks::*;
+
+mod utilities;
+use utilities::*;
+
+use console::network::Testnet3;
+use snarkvm_synthesizer::ConsensusStorage;
+
+use criterion::{BatchSize, Criterion};
+use std::fmt::Display;
+
+// Note: The number of commands that can be included in a finalize block must be within the range [1, 255].
+const NUM_COMMANDS: &[usize] = &[1, 2, 4, 8, 16, 32, 64, 128, 255];
+const NUM_EXECUTIONS: &[usize] = &[2, 4, 8, 16, 32, 64];
+const NUM_PROGRAMS: &[usize] = &[2, 4, 8, 16, 32, 64];
+
+/// A helper function for benchmarking `VM::add_next_block`.
+#[cfg(feature = "testing")]
+#[allow(unused)]
+pub fn bench_add_next_block<C: ConsensusStorage<Testnet3>>(
+    c: &mut Criterion,
+    header: impl Display,
+    mut workload: Workload,
+) {
+    // Setup the workload.
+    let (vm, private_key, benchmark_transactions, mut rng) = workload.setup::<C>();
+
+    // Run the benchmarks
+    for (name, transactions) in benchmark_transactions {
+        if transactions.is_empty() {
+            println!("Skipping benchmark {} because it has no transactions.", name);
+            continue;
+        }
+
+        let mut num_transactions = 0f64;
+        let mut num_rejected = 0f64;
+
+        c.bench_function(&format!("{header}/{name}/add_next_block"), |b| {
+            b.iter_batched(
+                // TODO: Do the blocks need to be removed on each iteration?
+                || {
+                    let block = construct_next_block(&vm, &private_key, &transactions, &mut rng).unwrap();
+                    num_transactions += block.transactions().iter().count() as f64;
```

Comment:
side note: `transactions` could probably easily return `impl ExactSizeIterator` instead of `impl Iterator`